src/libraries/Common/src/System/HexConverter.cs

```diff
@@ -0,0 +1,51 @@
+namespace System
+{
+    internal static class HexConverter
+    {
+        public static char ToCharUpper(int value)
+        {
+            value &= 0xf;
+            return (char)(value > 9 ? value - 10 + 'A' : value + '0');
+        }
+
+        public static char ToCharLower(int value)
+        {
+            value &= 0xf;
+            return (char)(value > 9 ? value - 10 + 'a' : value + '0');
+        }
+
+#if NETCOREAPP || NETSTANDARD2_1
+        public static unsafe string ToStringLower(ReadOnlySpan<byte> bytes)
+        {
+            fixed (byte* bytesPtr = bytes)
+            {
+                return string.Create(bytes.Length * 2, (Ptr: (IntPtr)bytesPtr, bytes.Length), (chars, args) =>
+                {
+                    int p = 0;
+                    foreach (byte b in new ReadOnlySpan<byte>((byte*)args.Ptr, args.Length))
+                    {
+                        chars[p++] = ToCharLower(b >> 4);
+                        chars[p++] = ToCharLower(b);
```

Comment:
This introduces quite a lot of bound checks. It may be better to iterate over `chars` (ideally by for-loop) to eliminate the bound checks for `chars`. For `b` you can operate on the pointer `bytesPtr` itself, no need to create a ROS for this.

With the other suggestions (especially) from @GrabYourPitchforks the creation could become completely branch-free.

Reply:
Also note that this could be vectorized (presumable not in this PR, but in a follow up (maybe I'll try something then)).